src/modules/vmount/vmount.cpp

```diff
@@ -173,278 +92,193 @@ static int vmount_thread_main(int argc, char *argv[])
-			output_config.yaw_offset = math::radians(params.mnt_off_yaw);
-			output_config.mavlink_sys_id_v1 = params.mnt_mav_sys_id_v1;
-			output_config.mavlink_comp_id_v1 = params.mnt_mav_comp_id_v1;
+	switch (params.mnt_mode_in) {
+	case 0:
+		// Automatic
+		// MAVLINK_V2 as well as RC input are supported together.
+		// Whichever signal is updated last, gets control, for RC there is a deadzone
+		// to avoid accidental activation.
+		thread_data.input_objs[thread_data.input_objs_len++] = new InputMavlinkGimbalV2(params);
 
-			bool alloc_failed = false;
-			thread_data.input_objs_len = 1;
+		thread_data.input_objs[thread_data.input_objs_len++] = new InputRC(params);
+		break;
 
-			if (test_input) {
-				thread_data.input_objs[0] = test_input;
+	case 1: // RC only
+		thread_data.input_objs[thread_data.input_objs_len++] = new InputRC(params);
+		break;
 
-			} else {
-				switch (params.mnt_mode_in) {
-				case 0:
-
-					// Automatic
-					thread_data.input_objs[0] = new InputMavlinkCmdMount();
-					thread_data.input_objs[1] = new InputMavlinkROI();
-
-					// RC is on purpose last here so that if there are any mavlink
-					// messages, they will take precedence over RC.
-					// This logic is done further below while update() is called.
-					thread_data.input_objs[2] = new InputRC(params.mnt_man_roll,
-										params.mnt_man_pitch,
-										params.mnt_man_yaw,
-										params.mnt_rate_pitch,
-										params.mnt_rate_yaw,
-										params.mnt_rc_in_mode);
-					thread_data.input_objs_len = 3;
+	case 2: // MAVLINK_ROI commands only (to be deprecated)
+		thread_data.input_objs[thread_data.input_objs_len++] = new InputMavlinkROI(params);
+		break;
 
-					break;
+	case 3: // MAVLINK_DO_MOUNT commands only (to be deprecated)
+		thread_data.input_objs[thread_data.input_objs_len++] = new InputMavlinkCmdMount(params);
+		break;
 
-				case 1: //RC
-					thread_data.input_objs[0] = new InputRC(params.mnt_man_roll,
-										params.mnt_man_pitch,
-										params.mnt_man_yaw,
-										params.mnt_rate_pitch,
-										params.mnt_rate_yaw,
-										params.mnt_rc_in_mode);
-					break;
+	case 4: //MAVLINK_V2
+		thread_data.input_objs[thread_data.input_objs_len++] = new InputMavlinkGimbalV2(params);
+		break;
 
-				case 2: //MAVLINK_ROI
-					thread_data.input_objs[0] = new InputMavlinkROI();
-					break;
+	default:
+		PX4_ERR("invalid input mode %" PRId32, params.mnt_mode_in);
+		break;
+	}
 
-				case 3: //MAVLINK_DO_MOUNT
-					thread_data.input_objs[0] = new InputMavlinkCmdMount();
-					break;
+	for (int i = 0; i < thread_data.input_objs_len; ++i) {
+		if (!thread_data.input_objs[i]) {
+			alloc_failed = true;
+		}
+	}
 
-				case 4: //MAVLINK_V2
-					thread_data.input_objs[0] = new InputMavlinkGimbalV2(
-						params.mav_sys_id,
-						params.mav_comp_id,
-						params.mnt_rate_pitch,
-						params.mnt_rate_yaw);
-					break;
+	if (alloc_failed) {
+		thread_data.input_objs_len = 0;
```

Comment:
Now there could be a leak if one of them is actually allocated.

Reply:
Thanks, let me check on that!

Reply:
I believe this is fixed with bca4f99fac.